File: src/code/cpp/dynamic_programming/kadane.cpp
#include <vector>
#include <algorithm>

using namespace std;


int kadane(vector<int>& arr) {
    int curr_sub = arr[0];
    int max_sub = arr[0];

    for (size_t i = 1; i < arr.size(); ++i) {
        curr_sub = max(curr_sub + arr[i], arr[i]);
        max_sub = max(max_sub, curr_sub);
    }

    return max_sub;
}

File: src/code/ruby/dynamic_programming/kadane.rb
def kadane(arr)
  curr_sub  = arr[0]
  max_sub   = arr[0]

  arr[1..-1].each do |num|
    curr_sub = [curr_sub + num, num].max
    max_sub = [max_sub, curr_sub].max
  end

  max_sub
end
